### PR TITLE
fix issue #41

### PR DESCRIPTION
**Pull request by lynus**
introduce isResourceAllocated() and support reconnect.

### CODE REVIEW

File: src/main/java/com/ibm/disni/RdmaActiveEndpointGroup.java [hunk 1]
```diff
@@ -110,9 +110,11 @@ public void close() throws IOException, InterruptedException {
...
 
 	void close(RdmaEndpoint endpoint) throws IOException {
-		IbvContext context = endpoint.getIdPriv().getVerbs();
-		RdmaActiveCqProcessor<C> cqProcessor = cqMap.get(context.getCmd_fd());
-		cqProcessor.unregister(endpoint);
+	    if (endpoint.isResouceAllocated()) {
+			IbvContext context = endpoint.getIdPriv().getVerbs();
+			RdmaActiveCqProcessor<C> cqProcessor = cqMap.get(context.getCmd_fd());
+			cqProcessor.unregister(endpoint);
+		}
```
[the remark this answers is not in the document]
Comment (lynus) [reply src/main/java/com/ibm/disni/RdmaActiveEndpointGroup.java]: My intention is not to prevent closing partial allocated endpoints, but to prevent releasing un-allocated resource when users decide to close the failed ep.

> Rather I suggest we check the allocation state in the endpoint and decide whether or not to de-allocate.

This is exactly what I am trying to achieve by introducing isResourceAllocated(). Application ep's close() implementation should be the following pattern:
`super.close();`
`if (isResourceAllocated())`
  `release its own resources...`

 Check if resources has been successfully allocated before releasing its own resources. super.close() can be called without checking is because current close() of RdmaEndpoint works fine with partial allocated ep, and RdmaActiveEndpoint's needs the patch you quoted above to work.

>  Also, this leaves partially allocated endpoints in the group, lingering around forever. 

This is not the case. Since ep throws exceptions before allocating resources https://github.com/zrlio/disni/blob/cfcee63cf054fb151f459357f601d7717f658b6c/src/main/java/com/ibm/disni/RdmaEndpoint.java#L116
this ep never registered into cqProcessor. It is also unregistered from group's clientEndpointMap in 
https://github.com/zrlio/disni/blob/cfcee63cf054fb151f459357f601d7717f658b6c/src/main/java/com/ibm/disni/RdmaEndpoint.java#L213
Therefore, no residues left after close().

In summary, users have to choices when catching exception on connect(): 1) close the failed ep and 2) try to connect() again.  I think the patch can properly preventing releasing un-allocated resources if users decide to take opt-1. And the logic in connect() can handle the opt-2 situation.
